Source/missiles.cpp

```diff
@@ -2804,7 +2732,9 @@ int AddMissile(Point src, Point dst, Direction midir, missile_id mitype, mienemy
 		PlaySfxLoc(missileData.mlSFX, missile.position.start);
 	}
 
+	pParentMissile = pParent;
 	missileData.mAddProc(missile, dst, midir);
+	pParentMissile = nullptr;
 
 	return mi;

```

Comment:
The only time the returned index is actually used in a lookup is in portal.cpp `AddWarpMissile`. If this function returns a pointer to the missile object you remove the need to store an ID with the missile and can remove the whole loop to find the first available ID. Can also change the data structure to a vector/linked list instead of a map.

Reply:
For devilutionX it would work. I was surprised that the missile id's aren't stored in other variables anymore. And we got rid of `dMissiles`.

But I hesitate to do this change:
- We would store pointers to `std::vector`. But these pointers are not stable if the vector is resized (elements removed or capacity is insufficient). For example your purposed `pcursmissile`. It's more fragile.
- Mod's can't use the id to store in other variables. For example if you want to add a mod feature to teleport to the last fired missile of a player.

On the other side we could simplify some logic.

Reply:
You'd need to use `std:::list` or `std::forward_list` for an STL container that doesn't invalidate iterators on resize/removal. The old array backed structure had the advantage of never moving a missile object once it had been allocated a space in the array (though it did change the order elements were accessed from the ActiveMissiles indirect lookup).

As for mods, if using a data structure where iterators aren't invalidated when adding/removing elements to the container they can still use a stored iterator at runtime. Saving and loading needs to translate from the iterator to the index using `std:::distance` but this is feasible with a bit of code isolated in the save/load functions.

Reply:
I don't think there is a need for preserving the IDs for potential mods. I would be more worried about how we would do a server/client based network protocol, but that is also just to fare removed to be a worry at this stage.

Reply:
I changed it to `std::list`. 😉